src/Microsoft.ML.StandardLearners/Standard/MultiClass/Pkpd.cs

```diff
@@ -68,94 +67,161 @@ public sealed class Pkpd : MetaMulticlassTrainer<PkpdPredictor, Pkpd.Arguments>
-            TDistPredictor[][] predictors;
-            predictors = new TDistPredictor[count][];
+            var predictors = new IPredictionTransformer<IPredictor>[count][];
+            var predModels = new TDistPredictor[count][];
+
             for (int i = 0; i < predictors.Length; i++)
             {
-                predictors[i] = new TDistPredictor[i + 1];
+                predictors[i] = new IPredictionTransformer<IPredictor>[i + 1];
+                predModels[i] = new TDistPredictor[i + 1];
+
                 for (int j = 0; j <= i; j++)
                 {
                     ch.Info($"Training learner ({i},{j})");
                     predictors[i][j] = TrainOne(ch, GetTrainer(), data, i, j);
+                    predModels[i][j] = (TDistPredictor)predictors[i][j].Model;
                 }
             }
-            return new PkpdPredictor(Host, predictors);
+
+            return new PkpdPredictor(Host, predModels);
         }
 
-        private TDistPredictor TrainOne(IChannel ch, TScalarTrainer trainer, RoleMappedData data, int cls1, int cls2)
+        private IPredictionTransformer<TDistPredictor> TrainOne(IChannel ch, TScalarTrainer trainer, RoleMappedData data, int cls1, int cls2)
         {
-            string dstName;
-            var view = MapLabels(data, cls1, cls2, out dstName);
-
-            var roles = data.Schema.GetColumnRoleNames()
-                .Where(kvp => kvp.Key.Value != CR.Label.Value)
-                .Prepend(CR.Label.Bind(dstName));
-            var td = new RoleMappedData(view, roles);
-
-            var predictor = trainer.Train(td);
-
-            ICalibratorTrainer calibrator;
-            if (Args.Calibrator == null)
-                calibrator = null;
-            else
-                calibrator = Args.Calibrator.CreateComponent(Host);
-            var res = CalibratorUtils.TrainCalibratorIfNeeded(Host, ch, calibrator, Args.MaxCalibrationExamples,
-                trainer, predictor, td);
-            var dist = res as TDistPredictor;
-            Host.Check(dist != null, "Calibrated predictor does not implement the expected interface");
-            Host.Check(dist is IValueMapperDist, "Calibrated predictor does not implement the IValueMapperDist interface");
-            return dist;
+            // this should not be necessary when the legacy constructor doesn't exist, and the label colum is not an optional parameter on the
+            // MetaMulticlassTrainer constructor.
+            string trainerLabel = data.Schema.Label.Name;
+
+            var view = MapLabels(data, cls1, cls2);
+            var transformer = trainer.Fit(view);
+
+            // the validations in the calibrator check for the feature column, in the RoleMappedData
+            var trainedData = new RoleMappedData(view, label: trainerLabel, feature: transformer.FeatureColumn);
+
+            var calibratedModel = transformer.Model as TDistPredictor;
```

Comment:
>calibratedModel [](start = 16, length = 15)

this logic is correct, replicate it to OVA, where it is different

Reply:
see my other question: the type for the model of the OVA IPredictionTransformer is TScalarPredictor. Double-checkign that it is correct to try casting to a TDistPredictor. Is that because learners that use calibration produce the propabilities and the score, therefore two floats?

---
In reply to: [217233894](https://github.com/dotnet/machinelearning/pull/865#discussion_r217233894) [](ancestors = 217233894)

Reply:
SOME TScalarPredictors are also TDistPredictors, some are not. If the one we have is not, we will need to calibrate. That's the gist of the code in lines 134-136

---
In reply to: [217477854](https://github.com/dotnet/machinelearning/pull/865#discussion_r217477854) [](ancestors = 217477854,217233894)